lib/puppet/util/character_encoding.rb

```diff
@@ -2,94 +2,88 @@
         # failure to encode!.
-        Puppet.debug(_("%{error}: %{value} is not valid UTF-8 and cannot be transcoded by Puppet.") %
+        Puppet.debug(_("%{error}: %{value} cannot be transcoded by Puppet.") %
           { error: detail.inspect, value: string.dump })
-        return nil
+        return string
       end
     end
 
-    private
+    # Given a string, tests if that string's bytes represent valid UTF-8, and if
+    # so return a copy of the string with external enocding set to UTF-8. Does
+    # not modify the byte representation of the string. If the string does not
+    # represent valid UTF-8, does not set the external encoding.
+    #
+    # This method is intended for situations where we do not believe that the
+    # encoding associated with a string is an accurate reflection of its actual
+    # bytes, i.e., effectively when we believe Ruby is incorrect in its
+    # assertion of the encoding of the string.
+    #
+    # @api public
+    # @param [String] string to set external encoding (re-label) to utf-8
+    # @return [String] a copy of string with external encoding set to utf-8, or
+    # the original string if override would result in invalid encoding.
+    def override_encoding_to_utf_8(string)
+      return string if string.encoding == Encoding::UTF_8
+      dup = string.dup
+      if dup.force_encoding(Encoding::UTF_8).valid_encoding?
+        return dup
+      else
+        Puppet.debug(_("%{value} is not valid UTF-8 and result of overriding encoding would be invalid.") % { value: string.dump })
+        return string
+      end
+    end
 
-    # Do our best to determine if a string is valid UTF-8 via String#valid_encoding? without permanently
-    # modifying or duplicating the string due to performance concerns
-    # @api private
-    # @param [String] string a string to test
-    # @return [Boolean] whether we think the string is UTF-8 or not
-    def valid_utf_8_bytes?(string)
-      original_encoding = string.encoding
-      valid = string.force_encoding(Encoding::UTF_8).valid_encoding?
-      string.force_encoding(original_encoding)
-      valid
+    # Given a string, return a copy of that string with any invalid byte
+    # sequences in its current encoding replaced with "?". We use "?" to make
+    # sure our output is consistent across ruby versions and encodings, and
+    # because calling scrub on a non-UTF8 string with the unicode replacement
+    # character "\uFFFD" results in an Encoding::CompatibilityError.
+    # @param string a string to remove invalid byte sequences from
+    # @return a copy of string invalid byte sequences replaced by "?" character
+    # @note does not modify encoding, but new string will have different bytes
+    #   from original. Only needed for ruby 1.9.3 support.
+    def scrub(string)
+      if String.method_defined?(:scrub)
+        string.scrub("?")
+      else
+        string.chars.map { |c| c.valid_encoding? ? c : "?" }.join
```

Comment:
And lets select the replacement character based on input string encoding here...

* `"\uFFFD"` for UTF-8
* `?` for everything else